Blockly Factory: Exporter Preview Code

quachtina96 opened the pull request:
# Show the user a preview of the code they are exporting.

### When they first visit the exporter tab, exporting block definitions and generator stubs are selected by default. 
The preview accordingly shows that these options are selected.
<img width="1437" alt="screen shot 2016-08-19 at 11 54 43 am" src="https://cloud.githubusercontent.com/assets/10423718/17822654/0ae63966-660d-11e6-87f2-efc0064c4987.png">

### The code preview updates when users select/deselect blocks by dragging them into the workspace.
<img width="1437" alt="screen shot 2016-08-19 at 9 51 27 am" src="https://cloud.githubusercontent.com/assets/10423718/17822640/00fb05e4-660d-11e6-81fa-139261b03d05.png">

### Unchecking block definitions or generator stubs in the export settings will remove the section accordingly in the preview.
<img width="1437" alt="screen shot 2016-08-19 at 11 55 29 am" src="https://cloud.githubusercontent.com/assets/10423718/17822658/12df35b4-660d-11e6-82bb-f9e9849383ef.png">

<img width="1439" alt="screen shot 2016-08-19 at 11 55 35 am" src="https://cloud.githubusercontent.com/assets/10423718/17822664/19ecfc88-660d-11e6-83b9-54f31def4aea.png">

# Code
- added divs and styling for the exporter preview
- added change listeners for the export settings to the app controller
- added functions for updating the preview in exporter controller

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/blockly/573)
<!-- Reviewable:end -->

quachtina96 commented:
@vicng @picklesrus @evd2014

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/573#-:-KPZQV7-H5ra5OFAPKDO:b-c2xsk1)*
<!-- Sent from Reviewable.io -->

evd2014 commented:
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[demos/blocklyfactory/app_controller.js, line 296 \[r2\]](https://reviewable.io:443/reviews/google/blockly/573#-KPnARHj_1FfzWuXwLp1:-KPnARHj_1FfzWuXwLp2:b-696e6g) ([raw file](https://github.com/google/blockly/blob/f343982b30f3a6c312457e74cde68233e43c9d09/demos/blocklyfactory/app_controller.js#L296)):*
> ```JavaScript
>     // Update toolbox to reflect current block library.
>     this.exporter.updateToolbox();
> ```

Should this also call updatePreview in the exporter in case the user has changed the definitions of blocks already defined in the block library?

---

*[demos/blocklyfactory/factory.css, line 217 \[r2\]](https://reviewable.io:443/reviews/google/blockly/573#-KPnAxvend6FXAegh-Cf:-KPnAxvfXhvQtlO3HsmF:brnmt23) ([raw file](https://github.com/google/blockly/blob/f343982b30f3a6c312457e74cde68233e43c9d09/demos/blocklyfactory/factory.css#L217)):*
> ```CSS
> }
> 
> #blockDefs .exportPreviewTextArea, #genStubs .exportPreviewTextArea {
> ```

Can remove #blockDefs and #genStubs, redundant because they are both of class .exportPreviewTextArea. Also remove the two .exportPreviewTextArea tags and just have one.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/573)*
<!-- Sent from Reviewable.io -->

quachtina96 commented:
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[demos/blocklyfactory/app_controller.js, line 296 \[r2\]](https://reviewable.io:443/reviews/google/blockly/573#-KPnARHj_1FfzWuXwLp1:-KPnO0ZRyPIFx0K3q9N6:b516xnn) ([raw file](https://github.com/google/blockly/blob/f343982b30f3a6c312457e74cde68233e43c9d09/demos/blocklyfactory/app_controller.js#L296)):*
<details><summary><i>Previously, evd2014 (Emma Dauterman) wrote…</i></summary>
> Should this also call updatePreview in the exporter in case the user has changed the definitions of blocks already defined in the block library?
</details>
Not needed in this version of the export selector (since the select/deselect handlers handle the block and leaving the exporter tab deselects all blocks). Still just added the call to updatePreview because I will need it with the new version of the block selector

---

*[demos/blocklyfactory/factory.css, line 217 \[r2\]](https://reviewable.io:443/reviews/google/blockly/573#-KPnAxvend6FXAegh-Cf:-KPnMg8pefCWe1TDEVdX:bjzmcav) ([raw file](https://github.com/google/blockly/blob/f343982b30f3a6c312457e74cde68233e43c9d09/demos/blocklyfactory/factory.css#L217)):*
<details><summary><i>Previously, evd2014 (Emma Dauterman) wrote…</i></summary>
> Can remove #blockDefs and #genStubs, redundant because they are both of class .exportPreviewTextArea. Also remove the two .exportPreviewTextArea tags and just have one.
</details>
good catch, thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/573)*
<!-- Sent from Reviewable.io -->

evd2014 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[demos/blocklyfactory/app_controller.js, line 296 \[r2\]](https://reviewable.io:443/reviews/google/blockly/573#-KPnARHj_1FfzWuXwLp1:-KPnTBNjUn707sSdA5sI:b-yfrsi9) ([raw file](https://github.com/google/blockly/blob/f343982b30f3a6c312457e74cde68233e43c9d09/demos/blocklyfactory/app_controller.js#L296)):*
<details><summary><i>Previously, quachtina96 (Tina Quach) wrote…</i></summary>
> Not needed in this version of the export selector (since the select/deselect handlers handle the block and leaving the exporter tab deselects all blocks). Still just added the call to updatePreview because I will need it with the new version of the block selector
</details>
Ok sounds good. Didn't know that leaving the exporter tab deselects all blocks (although that makes sense).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/573#-:-KPnTx2teszQKecLDP8T:bnfp4nl)*
<!-- Sent from Reviewable.io -->

picklesrus commented:
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[demos/blocklyfactory/app_controller.js, line 376 \[r1\]](https://reviewable.io:443/reviews/google/blockly/573#-KPZqtYAxjxIec6h9ZHu:-KPZqtYAxjxIec6h9ZHv:b-ayto55) ([raw file](https://github.com/google/blockly/blob/fa3fc44b1d9bea4fa9ca999104739cb8ba687c5c/demos/blocklyfactory/app_controller.js#L376)):*
> ```JavaScript
>   document.getElementById('blockDefCheck').addEventListener('change',
>       function(e) {
>         document.getElementById('blockDefs_label').style.display =
> ```

A bunch of these have the same form (i.e go check an element if an element is checked, then set display to block or none based on that). Refactoring that into a function might clean this up a little. 

---

*[demos/blocklyfactory/block_exporter_controller.js, line 336 \[r3\]](https://reviewable.io:443/reviews/google/blockly/573#-KPnkNv177KEKUcIFnmv:-KPnkNv177KEKUcIFnmw:b-vjzasb) ([raw file](https://github.com/google/blockly/blob/74b504db60f325be080aeeec26c66a27a0e913ad/demos/blocklyfactory/block_exporter_controller.js#L336)):*
> ```JavaScript
>  * selected by the user.
>  */
> BlockExporterController.prototype.updateBlockDefsFormat = function() {
> ```

these three functions are fairly repetitive - is there a way to refactor to share code between them? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/573)*
<!-- Sent from Reviewable.io -->

vicng commented:
+1 on Katelyn's comments

---

Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/573#-:-KPoMRaUsvnL8xfKWC17:b-w71s6e)*
<!-- Sent from Reviewable.io -->

quachtina96 commented:
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[demos/blocklyfactory/app_controller.js, line 376 \[r1\]](https://reviewable.io:443/reviews/google/blockly/573#-KPZqtYAxjxIec6h9ZHu:-KPovMLcdXsL5ORkSwVg:b-896fix) ([raw file](https://github.com/google/blockly/blob/fa3fc44b1d9bea4fa9ca999104739cb8ba687c5c/demos/blocklyfactory/app_controller.js#L376)):*
<details><summary><i>Previously, picklesrus wrote…</i></summary>
> A bunch of these have the same form (i.e go check an element if an element is checked, then set display to block or none based on that). Refactoring that into a function might clean this up a little. 
</details>
Done.

---

*[demos/blocklyfactory/block_exporter_controller.js, line 336 \[r3\]](https://reviewable.io:443/reviews/google/blockly/573#-KPnkNv177KEKUcIFnmv:-KPonZPpPgQZDW1fE8P7:b-896fix) ([raw file](https://github.com/google/blockly/blob/74b504db60f325be080aeeec26c66a27a0e913ad/demos/blocklyfactory/block_exporter_controller.js#L336)):*
<details><summary><i>Previously, picklesrus wrote…</i></summary>
> these three functions are fairly repetitive - is there a way to refactor to share code between them? 
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/573)*
<!-- Sent from Reviewable.io -->

vicng commented:
Review status: 0 of 6 files reviewed at latest revision, 6 unresolved discussions.

---

*[demos/blocklyfactory/block_exporter_controller.js, line 327 \[r4\]](https://reviewable.io:443/reviews/google/blockly/573#-KPoyBk5sJAfGRvUYNPB:-KPoyBk5sJAfGRvUYNPC:bhld48l) ([raw file](https://github.com/google/blockly/blob/3f0bbdbd50d0feb7f49f01561fa2f71df157d37c/demos/blocklyfactory/block_exporter_controller.js#L327)):*
> ```JavaScript
>  * Updates block definitions in the exporter preview.
>  */
> BlockExporterController.prototype.updateBlockDefs = function() {
> ```

It's bad practice to abbreviate method names.

Does this method need to exist? Could you just call updatePreview()?

---

*[demos/blocklyfactory/block_exporter_controller.js, line 337 \[r4\]](https://reviewable.io:443/reviews/google/blockly/573#-KPoyW3zgmSn8EcBVr5z:-KPoyW3zgmSn8EcBVr6-:b2zwedf) ([raw file](https://github.com/google/blockly/blob/3f0bbdbd50d0feb7f49f01561fa2f71df157d37c/demos/blocklyfactory/block_exporter_controller.js#L337)):*
> ```JavaScript
>  * Updates generator stubs in the exporter preview.
>  */
> BlockExporterController.prototype.updateGenStubs = function() {
> ```

Same thing here too.

---

*[demos/blocklyfactory/block_exporter_controller.js, line 361 \[r4\]](https://reviewable.io:443/reviews/google/blockly/573#-KPoyXnr543oYmFrJPa3:-KPoyXnr543oYmFrJPa4:b-qvuyzw) ([raw file](https://github.com/google/blockly/blob/3f0bbdbd50d0feb7f49f01561fa2f71df157d37c/demos/blocklyfactory/block_exporter_controller.js#L361)):*
> ```JavaScript
>  * in the format specified in export settings.
>  */
> BlockExporterController.prototype.getBlockDefsOfSelected = function() {
> ```

Fix method name.

---

*[demos/blocklyfactory/block_exporter_controller.js, line 376 \[r4\]](https://reviewable.io:443/reviews/google/blockly/573#-KPoyZUCNcIjbZxkmb5k:-KPoyZUCNcIjbZxkmb5l:b-k2pczf) ([raw file](https://github.com/google/blockly/blob/3f0bbdbd50d0feb7f49f01561fa2f71df157d37c/demos/blocklyfactory/block_exporter_controller.js#L376)):*
> ```JavaScript
>  * in the language specified in export settings.
>  */
> BlockExporterController.prototype.getGenStubsOfSelected = function() {
> ```

Here too.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/573)*
<!-- Sent from Reviewable.io -->

quachtina96 commented:
Review status: 0 of 6 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[demos/blocklyfactory/block_exporter_controller.js, line 327 \[r4\]](https://reviewable.io:443/reviews/google/blockly/573#-KPoyBk5sJAfGRvUYNPB:-KPp1nJvezXd4AD9LIta:bpjggxt) ([raw file](https://github.com/google/blockly/blob/3f0bbdbd50d0feb7f49f01561fa2f71df157d37c/demos/blocklyfactory/block_exporter_controller.js#L327)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> It's bad practice to abbreviate method names.
> 
> Does this method need to exist? Could you just call updatePreview()?
</details>
Removed this func

---

*[demos/blocklyfactory/block_exporter_controller.js, line 337 \[r4\]](https://reviewable.io:443/reviews/google/blockly/573#-KPoyW3zgmSn8EcBVr5z:-KPp1pMiwKkAQy3XiXxD:bvfjvfl) ([raw file](https://github.com/google/blockly/blob/3f0bbdbd50d0feb7f49f01561fa2f71df157d37c/demos/blocklyfactory/block_exporter_controller.js#L337)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> Same thing here too.
</details>
removed this func

---

*[demos/blocklyfactory/block_exporter_controller.js, line 361 \[r4\]](https://reviewable.io:443/reviews/google/blockly/573#-KPoyXnr543oYmFrJPa3:-KPp1rewa9oMKQreZXeV:bb7f2mf) ([raw file](https://github.com/google/blockly/blob/3f0bbdbd50d0feb7f49f01561fa2f71df157d37c/demos/blocklyfactory/block_exporter_controller.js#L361)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> Fix method name.
</details>
done

---

*[demos/blocklyfactory/block_exporter_controller.js, line 376 \[r4\]](https://reviewable.io:443/reviews/google/blockly/573#-KPoyZUCNcIjbZxkmb5k:-KPp1sN0oiG--Wva5AS-:bb7f2mf) ([raw file](https://github.com/google/blockly/blob/3f0bbdbd50d0feb7f49f01561fa2f71df157d37c/demos/blocklyfactory/block_exporter_controller.js#L376)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> Here too.
</details>
done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/573)*
<!-- Sent from Reviewable.io -->

vicng commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 7 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/573#-:-KPp4LKn-_lHs0te_-E6:bnfp4nl)*
<!-- Sent from Reviewable.io -->
